crates/sol-types/src/types/encodable.rs

```diff
@@ -0,0 +1,405 @@
+    ///
+    /// See [`SolType::sol_type_name`] for more information.
+    #[inline]
+    fn sol_type_name(&self) -> Cow<'static, str> {
+        Self::SolType::sol_type_name()
+    }
+
+    /// Calculate the ABI-encoded size of the data.
+    ///
+    /// See [`SolType::abi_encoded_size`] for more information.
+    #[inline]
+    fn abi_encoded_size(&self) -> usize
+    where
+        Self: SolTypeEncodable<Self::SolType>,
+    {
+        <Self as SolTypeEncodable<Self::SolType>>::abi_encoded_size(self)
+    }
+
+    /// Encode this data according to EIP-712 `encodeData` rules, and hash it
+    /// if necessary.
+    ///
+    /// See [`SolType::eip712_data_word`] for more information.
+    #[inline]
+    fn eip712_data_word(&self) -> Word
+    where
+        Self: SolTypeEncodable<Self::SolType>,
+    {
+        <Self as SolTypeEncodable<Self::SolType>>::eip712_data_word(self)
+    }
+
+    /// Non-standard Packed Mode ABI encoding.
+    ///
+    /// See [`SolType::abi_encode_packed_to`] for more information.
+    #[inline]
+    fn abi_encode_packed_to(&self, out: &mut Vec<u8>)
+    where
+        Self: SolTypeEncodable<Self::SolType>,
+    {
+        <Self as SolTypeEncodable<Self::SolType>>::abi_encode_packed_to(self, out)
+    }
+
+    /// Non-standard Packed Mode ABI encoding.
+    ///
+    /// See [`SolType::abi_encode_packed`] for more information.
+    #[inline]
+    fn abi_encode_packed(&self) -> Vec<u8>
+    where
+        Self: SolTypeEncodable<Self::SolType>,
+    {
+        let mut out = Vec::new();
+        <Self as SolTypeEncodable<Self::SolType>>::abi_encode_packed_to(self, &mut out);
+        out
+    }
+
+    /// ABI-encodes the value.
+    ///
+    /// See [`SolType::abi_encode`] for more information.
+    #[inline]
+    fn abi_encode(&self) -> Vec<u8>
+    where
+        Self: SolTypeEncodable<Self::SolType>,
+    {
+        Self::SolType::abi_encode(self)
+    }
+
+    /// Encodes an ABI sequence.
+    ///
+    /// See [`SolType::abi_encode_sequence`] for more information.
+    #[inline]
+    fn abi_encode_sequence(&self) -> Vec<u8>
+    where
+        Self: SolTypeEncodable<Self::SolType>,
+        for<'a> <Self::SolType as SolType>::TokenType<'a>: TokenSeq<'a>,
+    {
+        Self::SolType::abi_encode_sequence(self)
+    }
+
+    /// Encodes an ABI sequence suitable for function parameters.
+    ///
+    /// See [`SolType::abi_encode_params`] for more information.
+    #[inline]
+    fn abi_encode_params(&self) -> Vec<u8>
+    where
+        Self: SolTypeEncodable<Self::SolType>,
+        for<'a> <Self::SolType as SolType>::TokenType<'a>: TokenSeq<'a>,
+    {
+        Self::SolType::abi_encode_params(self)
+    }
+
+    /// ABI-decode this type from the given data.
+    ///
+    /// See [`SolType::abi_decode`] for more information.
+    fn abi_decode(data: &[u8], validate: bool) -> Result<<Self::SolType as SolType>::RustType> {
+        Self::SolType::abi_decode(data, validate)
+    }
```

Comment:
I don't think there's a way to say `where Self = Self::SolType::RustType` so for example `str::abi_decode` will return `String` here, see tests below

Reply:
what about using from? 

`where Self: From<Self::SolType::RustType>`